File: dotnet/private/tests/resources/EmbeddedResource.fs
module Program

let private assertEqual<'T when 'T : equality> (expected: 'T) (actual: 'T) =
    if not (expected.Equals(actual)) then
        raise (System.Exception(sprintf "Expected %A but got %A" expected actual))

let private assertContains<'T when 'T : equality> (haystack: seq<'T>) (needle: 'T) =
    if not (Seq.contains needle haystack) then
        let itemsStr = haystack |> Seq.map (fun x -> x.ToString()) |> String.concat ", "
        raise (System.Exception(sprintf "Expected %A to be in [%s]" needle itemsStr))

[<EntryPoint>]
let main _ =
    let assembly = System.Reflection.Assembly.GetExecutingAssembly()
    let resources = assembly.GetManifestResourceNames()
    assertContains resources "EmbeddedResource.Library.nested.path.to.resource.txt"

    use stream = assembly.GetManifestResourceStream(
        "EmbeddedResource.Library.nested.path.to.resource.txt"
    )
    use reader = new System.IO.StreamReader(stream)
    let content = reader.ReadToEnd().Trim()
    assertEqual "Well hello friends! :^)" content
    0

File: dotnet/private/tests/resources/EmbeddedResource.cs
using System.Linq;

public static class Program
{
    private static void AssertEqual<T>(T expected, T actual)
    {
        if (!expected.Equals(actual))
        {
            throw new System.Exception($"Expected {expected} but got {actual}");
        }
    }

    private static void AssertContains<T>(
        System.Collections.Generic.IEnumerable<T> haystack,
        T needle
    )
    {
        if (!haystack.Contains(needle))
        {
            throw new System.Exception(
                $"Expected {needle} to be in [{string.Join(", ", haystack.Select(x => x.ToString()))}]"
            );
        }
    }

    public static void Main()
    {
        var assembly = System.Reflection.Assembly.GetExecutingAssembly();
        var resources = assembly.GetManifestResourceNames();
        AssertContains(resources, "EmbeddedResource.Library.nested.path.to.resource.txt");

        using var stream = assembly.GetManifestResourceStream(
            "EmbeddedResource.Library.nested.path.to.resource.txt"
        );
        using var reader = new System.IO.StreamReader(stream);
        var content = reader.ReadToEnd();
        AssertEqual("Well hello friends! :^)", content.Trim());
    }
}
